drivers/gpio/gpio_sim.c

```diff
@@ -0,0 +1,563 @@
+}
+
+/*
+ * GPIO backend API (for setting input pin values)
+ */
+
+static void gpio_sim_pend_interrupt(struct device *port, gpio_port_pins_t mask,
+				    gpio_port_value_t prev_values,
+				    gpio_port_value_t values)
+{
+	size_t i;
+	struct gpio_sim_data *drv_data =
+		(struct gpio_sim_data *)port->driver_data;
+	struct gpio_sim_config *config =
+		(struct gpio_sim_config *)drv_data->config;
+	gpio_port_pins_t interrupts = 0;
+
+	for (i = 0; mask && i < CONFIG_GPIO_SIM_NPINS_PER_PORT;
+	     ++i, mask >>= 1, prev_values >>= 1, values >>= 1) {
+		if (!(mask & 1)) {
+			continue;
+		}
+
+		bool prev_bit = prev_values & 1;
+		bool bit = values & 1;
+
+		switch (config->flags[i] & interrupt_bitmask) {
+		case GPIO_INT_EDGE_RISING:
+			if (!prev_bit && bit) {
+				interrupts |= (1 << i);
+			}
+			break;
+
+		case GPIO_INT_EDGE_FALLING:
+			if (prev_bit && !bit) {
+				interrupts |= (1 << i);
+			}
+			break;
+
+		case GPIO_INT_EDGE_BOTH:
+			if (prev_bit != bit) {
+				interrupts |= (1 << i);
+			}
+			break;
+
+		case GPIO_INT_LEVEL_LOW:
+			if (!bit) {
+				interrupts |= (1 << i);
+			}
+			break;
+
+		case GPIO_INT_LEVEL_HIGH:
+			if (bit) {
+				interrupts |= (1 << i);
+			}
+			break;
+
+		case 0:
+		case GPIO_INT_DISABLE:
+			break;
+
+		default:
+			LOG_DBG("unhandled case %u",
+			  config->flags[i] & interrupt_bitmask);
+			break;
+		}
+	}
+
+	config->interrupts |= interrupts;
+
+	if (interrupts) {
+		gpio_fire_callbacks(&config->callbacks, port, interrupts);
+	}
+}
+
+int gpio_sim_input_set_masked(struct device *port, gpio_port_pins_t mask,
+			      gpio_port_value_t values)
+{
+	int r;
+	struct gpio_sim_data *drv_data =
+		(struct gpio_sim_data *)port->driver_data;
+	struct gpio_sim_config *config =
+		(struct gpio_sim_config *)drv_data->config;
+	gpio_port_pins_t input_mask;
+	gpio_port_pins_t prev_values;
+
+	if (mask == 0) {
+		r = 0;
+		goto out;
+	}
+
+	k_mutex_lock(&config->mu, K_FOREVER);
+
+	if (~config->common.port_pin_mask & mask) {
+		r = EINVAL;
+		goto unlock;
+	}
+
+	if (values & ~mask) {
+		r = EINVAL;
+		goto unlock;
+	}
+
+	input_mask = gpio_sim_input_mask(config);
+
+	if (~input_mask & mask) {
+		r = EINVAL;
+		goto unlock;
+	}
+
+	prev_values = config->vals;
+
+	config->vals &= ~mask;
+	config->vals |= values;
+
+	values = config->vals;
+
+	gpio_sim_pend_interrupt(port, mask, prev_values, values);
+
+	r = 0;
+
+unlock:
+	k_mutex_unlock(&config->mu);
+
+out:
+	return r;
+}
+
+int gpio_sim_output_get_masked(struct device *port, gpio_port_pins_t mask,
+			       gpio_port_value_t *values)
+{
+	int r;
+	struct gpio_sim_data *drv_data =
+		(struct gpio_sim_data *)port->driver_data;
+	struct gpio_sim_config *config =
+		(struct gpio_sim_config *)drv_data->config;
+	gpio_port_pins_t output_mask;
+
+	if (mask == 0) {
+		r = 0;
+		goto out;
+	}
+
+	k_mutex_lock(&config->mu, K_FOREVER);
+
+	if (~config->common.port_pin_mask & mask) {
+		r = EINVAL;
+		goto unlock;
+	}
+
+	output_mask = gpio_sim_output_mask(config);
+
```

Comment:
In general it feels like there are two many blank lines. This line seems linked to the if() below, so I think it should be together, for example

Reply:
Fixed.